### PR TITLE
[TOPSORT] Action destination for Topsort Events

### PR DESCRIPTION
**Pull request by agustinespildora**
# Topsort Events Actions Destination

This is a destination so that eCommerce sites that already use Segment (to track actions) can integrate Topsort events for sponsored listings seamlessly. 

## Intro to Topsort Events
Topsort allows to monetize through auction ads served in your site while respecting users privacy and helping sellers get the most out of their ads increasing conversions. To learn more about Topsort you can visit our [docs](https://docs2.topsort.com/).
This destination will help you track the 3 main events Topsort needs to serve the most relevant ads and accomplish attribution for sponsored listings: `impressions`, `clicks` and `purchases`. Here is an explanation of what each one of these means to Topsort:

1. **Impression**: Every time a user sees a promoted product in a listing page or section.
2. **Click**: If the user then clicks or adds the promoted product to a cart, then a `click` should be triggered.
3. **Purchase**: Finally, when an order is completed and confirmed to be successful, a `purchase` event should be reported to Topsort. You can report a purchase whether it has promoted products or not. We will take care of filtering relevant promoted products inside the completed order given the information about promoted clicks.

This destination has 3 presets that map the Track events `'Product Viewed'`, `'Product Clicked'` and `'Order Completed'` to Topsort's `impression`, `click` and `purchase` events respectively. You can find more about the different fields that each of these events allow in the [Topsort Events API](https://docs2.topsort.com/api/auctions-events-v2/operations/reportevents/).

## Future improvements
In the future we will map other Segment events to these 3 Topsort events, like `'Product List Viewed'`, `'AddToCart'` and others.

## Testing
Added tests for success and failing when a required field is not passed, for every action.

- [x] Added required fields
- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Add Auth test
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)

#### Purchase
image

#### Click
image

#### Impression
image

### CODE REVIEW

**Review by agustinespildora — COMMENTED** [review]
To do:
- [ ] ~~ add limits to numeric values~~ (not necessary as values from segment should be greater than 0)
- [x] fix partner actions

**Comment by joe-ayoub-segment** [comment]
Thanks for raising this PR @agustinespildora . 
Either myself of my colleage @tcgilbert will review the PR once you have marked it as ready for review.

File: packages/destination-actions/src/destinations/topsort/impression/index.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import type { ActionDefinition } from '@segment/actions-core'
+import type { Settings } from '../generated-types'
+import type { Payload } from './generated-types'
+import { TopsortAPIClient } from '../client'
+
+const action: ActionDefinition<Settings, Payload> = {
+  title: 'Impression',
+  defaultSubscription: 'type = "track" and event = "Product Viewed"',
+  description: 'A set of impressions means such promotables have become visible to the consumer.',
```
Comment (tcgilbert) [inline packages/destination-actions/src/destinations/topsort/impression/index.ts:9]:
```suggestion
description: 'Send impression events to Topsort when a consumer has viewed a promotable.'
```

File: packages/destination-actions/src/destinations/topsort/purchase/index.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
+import type { ActionDefinition } from '@segment/actions-core'
+import type { Settings } from '../generated-types'
+import type { Payload } from './generated-types'
+import { TopsortAPIClient } from '../client'
+
+const action: ActionDefinition<Settings, Payload> = {
+  title: 'Purchase',
+  defaultSubscription: 'type = "track" and event = "Order Completed"',
+  description: 'A purchase is sent to Topsort once a consumer places an order.',
```
Comment (tcgilbert) [inline packages/destination-actions/src/destinations/topsort/purchase/index.ts:9]:
```suggestion
  description: 'Send purchase events to Topsort when a consumer places an order.',
```

**Review by tcgilbert — CHANGES_REQUESTED** [review]

**Comment by tcgilbert** [comment]
@agustinespildora thanks for the thorough PR!  just some minor updates to the action descriptions. 

Additionally, the unit test CI is failing. Looks like it is one of the snapshots being outdated. Can you run the command below, and commit the changes? 

```
yarn jest --testPathPattern='./packages/destination-actions/src/destinations/topsort' --updateSnapshot 
```

**Comment by agustinespildora** [comment]
> @agustinespildora thanks for the thorough PR! just some minor updates to the action descriptions.
> 
> Additionally, the unit test CI is failing. Looks like it is one of the snapshots being outdated. Can you run the command below, and commit the changes?
> 
> ```
> yarn jest --testPathPattern='./packages/destination-actions/src/destinations/topsort' --updateSnapshot 
> ```

Thanks for the quick review @tcgilbert! I updated the snapshots, let me know if anyhting else is needed.

File: packages/destination-actions/src/destinations/topsort/impression/index.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
+import type { ActionDefinition } from '@segment/actions-core'
+import type { Settings } from '../generated-types'
+import type { Payload } from './generated-types'
+import { TopsortAPIClient } from '../client'
+
+const action: ActionDefinition<Settings, Payload> = {
+  title: 'Impression',
+  defaultSubscription: 'type = "track" and event = "Product Viewed"',
+description: 'Send impression events to Topsort when a consumer has viewed a promotable.'
```
Comment (tcgilbert) [inline packages/destination-actions/src/destinations/topsort/impression/index.ts:9]:
```suggestion
description: 'Send impression events to Topsort when a consumer has viewed a promotable.',
```

**Review by tcgilbert — CHANGES_REQUESTED** [review]
one more change

**Comment by agustinespildora** [comment]
Hey @tcgilbert corrected the missing comma.
Can you help me figure out how can I fix the snyk CI run that failed?

**Review by tcgilbert — APPROVED** [review]

**Comment by tcgilbert** [comment]
> Hey @tcgilbert corrected the missing comma. Can you help me figure out how can I fix the snyk CI run that failed?

Snyk is expected to fail for external PRs, so no worries there. 

I've approved the pull request, and we should be able to get this included in the next deploy on Monday. We'll get back to you once that happens, Meanwhile, if you have any questions or concerns, do let me know.

**Review by joe-ayoub-segment — APPROVED** [review]
I didn't review - approving as Thomas labeled ready for release.

**Comment by joe-ayoub-segment** [comment]
hi @agustinespildora and @tcgilbert this PR has been deployed.

**Pull request merged**